slice-core/src/main/java/com/cognifide/slice/core/internal/monitoring/InjectorStatisticsRepositoryImpl.java

```diff
@@ -0,0 +1,63 @@
+/*-
+ * #%L
+ * Slice - Core
+ * %%
+ * Copyright (C) 2012 Cognifide Limited
+ * %%
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+package com.cognifide.slice.core.internal.monitoring;
+
+import java.util.Queue;
+
+import com.cognifide.slice.api.execution.ExecutionContext;
+import com.google.inject.Singleton;
+
+@Singleton
+public class InjectorStatisticsRepositoryImpl implements InjectorStatisticsRepository {
+
+	private final ModelUsageData modelUsageDataRoot = new ModelUsageData();
+
+	@Override
+	public InjectionMonitoringContext startMonitoring() {
+		return new InjectionMonitoringContext();
+	}
+
+	@Override
+	public void save(InjectionMonitoringContext ctx) {
+		saveEvent(ctx.getInjecteeClass(), ctx.getElapsedTime(), ctx.getModelsStack());
+	}
+
+	private void saveEvent(Class<?> type, Long timeMeasurement, Queue<ExecutionContext> modelHierarchyContext) {
+		ModelUsageData properItemInHierarchy = modelUsageDataRoot;
+		while (modelHierarchyContext.peek() != null) {
+			Class<?> ctx = modelHierarchyContext.poll().getInjecteeClass();
+			synchronized (this) {

```

Comment:
I would rather challenge this 'synchronized' block. I don't think it is really needed... provided that we will use putIfAbsent() instead of regular put.
What would be the worst scenario then?
1. If gives us true - there is no such a key yet
2. We are putting new MUD (using putIfAbsent)
3. Between 'if' and 'putIfAbsent' there were another thread that put its own MUD object
4. As a result get(ctx) few lines later returns the object from that other thread instead of ours - which is absolutely fine, because our object was still empty and wasn't carrying any information

Because of that I suggest to simply remove synchronized and introduce putIfAbsent.